.github/workflows/preview-docs.yml

```
@@ -1,28 +1,88 @@
 name: Preview docs
 
 on:
-  pull_request_target:
+  workflow_run:
+    workflows: [docs]
+    types: [requested, completed]
 
 jobs:
+  pre-deploy:
+    if: github.repository == 'mlflow/mlflow' && github.event.workflow_run.status == 'requested'
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: write # preview_docs.py comments on PRs
+    timeout-minutes: 10
+    steps:
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+      - uses: ./.github/actions/setup-python
+      - name: Post placeholder comment
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR_NUMBER: ${{ github.event.workflow_run.pull_requests[0].number }}
+        run: |
+          python dev/preview_docs.py \
+            --commit-sha ${{ github.event.workflow_run.head_sha }} \
+            --pull-number $PR_NUMBER \
+            --workflow-run-id ${{ github.run_id }} \
+            --docs-workflow-run-url ${{ github.event.workflow_run.html_url }} \
+            --stage requested
+
   main:
-    if: github.repository == 'mlflow/mlflow'
+    if: github.repository == 'mlflow/mlflow' && github.event.workflow_run.status == 'completed'
     runs-on: ubuntu-latest
     permissions:
-      statuses: read # preview_docs.py checks PR statuses
       pull-requests: write # preview_docs.py comments on PRs
     timeout-minutes: 10
     steps:
+      - name: Handle docs workflow failure
+        if: github.event.workflow_run.conclusion == 'failure'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR_NUMBER: ${{ github.event.workflow_run.pull_requests[0].number }}
+        run: |
+          python dev/preview_docs.py \
+            --commit-sha ${{ github.event.workflow_run.head_sha }} \
+            --pull-number $PR_NUMBER \
+            --workflow-run-id ${{ github.run_id }} \
+            --docs-workflow-run-url ${{ github.event.workflow_run.html_url }} \
+            --stage failed
+          exit 0 # Exit the job if the docs workflow failed
       - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
       - uses: ./.github/actions/setup-python
+      - uses: ./.github/actions/setup-node
+      - name: Download build artifact
+        uses: actions/download-artifact@95815c38cf2ff2164869cbab79da8d1f422bc89e # v4.2.1
+        with:
+          name: website-build-${{ github.event.workflow_run.id }}
```

```suggestion
          name: docs-build-${{ github.event.workflow_run.id }}
```

nit: can we rename this?